third_party/amd/lib/TritonAMDGPUTransforms/AMDBypassLDSForDotOperand.cpp

```diff
@@ -0,0 +1,185 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// Coalescing in Triton:
+//
+// Coalescing in Triton is managed by configuring parameters for the blocked
+// layout during the Coalesce pass. There are two primary levels of
+// coalescing to consider:
+//
+// 1. Maximizing Load Width:
+//    Achieving the widest possible loads ensures that elements are grouped into
+//    larger memory transactions by the VMEM unit. This reduces the number of
+//    instructions needed to load data, minimizing instruction queue size and
+//    reducing wait times.
+//
+// 2. Ensuring Consecutive Thread Access:
+//    When consecutive threads access sequential memory addresses, the TA unit
+//    in the VMEM can merge multiple requests into a single, larger transaction.
+//    This approach significantly boosts memory bandwidth utilization.
+//
+//===----------------------------------------------------------------------===//
+//
+// Optimizing the Data Flow:
+//
+// Under certain conditions, the dot layout of one of the operands allows direct
+// loading from HBM to VGPRs in the MFMA dot layout, without losing level 1
+// coalescing efficiency or increasing the number of global loads due to shared
+// data between threads.
+//
+// The required conditions are:
+//
+// 1. K-Contig (K dimension is continuous) Tensor Layout :
+//    The operand we want to bypass LDS for must be K-contig (i.e., row-major
+//    for operand 0 or column-major for operand 1). This supports vectorized
+//    global load instructions, as MFMA instructions require each thread to hold
+//    B operand elements along the K dimension.
+//
+// 2. kWidth * sizeof(dataType) == 128:
+//    Using the maximum kWidth for a specific data type ensures optimal global
+//    load vectorization (e.g., using global_load_dwordx4 instructions).
+//
+// 3. Single Warp per CTA Dimension:
+//    Either warpsPerCTA[ndim] == 1 for operand A bypass or warpsPerCTA[mDim] ==
+//    1 for operand B bypass. This guarantees that each tensor element is
+//    handled by exactly one thread, maintaining the same number of global loads
+//    as in the blocked layout (i.e., each element is loaded only once).
+//
+//===----------------------------------------------------------------------===//
+// Current Limitations:
+// These limitations are temporary and will be addressed in future updates:
+//
+// 1. Support is limited to bypassing LDS for operand 1 (e.g., weights in
+//    MoE-like kernels). Bypassing for operand 0 is not yet implemented.
+//===----------------------------------------------------------------------===//
+
+using namespace mlir;
+namespace ttg = triton::gpu;
+
+// Find all tt.load instructions that are involved in computation of a tensor
+// for operand that is getting converted to dot layout.
+SmallVector<triton::LoadOp> getAllLoadOpsReachingOp(Operation *op) {
+  SmallVector<triton::LoadOp> loadOpsVec;
+  SetVector<Operation *> visited;
```

Comment:
We just need a set here? No need to be `SetVector`?

I guess another way is to use `getBackwardSlice` and then do a second round of filtering from the result to avoid doing the manual worklist here.